manifests/charts/gateways/istio-egress/templates/zzz_profile.yaml

```diff
@@ -27,12 +58,37 @@ Finally, we can set all of that under .Values so the chart behaves without aware
 {{- end }}
 {{- with .Values.compatibilityVersion }}
 {{- with $.Files.Get (printf "files/profile-compatibility-version-%s.yaml" .) }}
-{{- $ignore := mustMergeOverwrite $profile (. | fromYaml) }}
+{{- $ignore := mustMergeOverwrite $profile (. | fromYaml ) }}
 {{- else }}
 {{ fail (cat "unknown compatibility version" $.Values.compatibilityVersion) }}
 {{- end }}
 {{- end }}
 {{- if $profile }}
 {{- $a := mustMergeOverwrite $defaults $profile }}
 {{- end }}
+
 {{- $b := set $ "Values" (mustMergeOverwrite $defaults $.Values) }}
+
+
+{{/*
```

Comment:
I am not sure why we need all of this. Does https://github.com/istio/istio/pull/51749 meet the goals you are trying to?

Reply:
Yep #51749 is a subset of the changes currently in this PR (`istioctl` squish fix + handle globals in a reasonable way).

This PR is now doing some additional things, like fixing the legacy `pilot/cni` scoping as well, making some test fixes, etc.

Happy to merge #51749 first to descope this PR, then rebase + merge the fixes here on top, if you'd prefer.

Reply:
I'm just not quite sure why we want to change the pilot and cni charts. They are pretty straightforward and simple and I haven't heard any complaints? I worry we are overcomplicating things when the real goal was just to get globals into ztunnel

Reply:
We have gotten (what I consider pretty reasonable) complaints that for 2 of our many charts you have to set `cni.hub/pilot.hub` etc, when for all the others you can just do `hub/tag` - again, it's a consistency problem, which I don't think we actually need here.

This also cuts against Helm conventions a bit - if you compose those charts with different behavior, you get values paths like `cni.cni.hub` which is just odd if the chart has exactly one component in it (which it does, and which I think we have mentioned before is the invariant we are sticking with).

The bigger problem tho is that what you have in #51749 can't really work consistently as-is, as mentioned in https://github.com/istio/istio/pull/51749/files#r1657683855 - we have to do a bit more normalization to get consistent behavior.

Reply:
Yes, the legacy charts have legacy semantics. Please don't change these without getting a clear understanding of our goals. I very much don't like these changes without clear high level direction + mixed in with "ztunnel not respecting global".

We are in the `cni.cni.hub` position for very similar reasons (caused by me). We went from "Istio is only a parent chart" to "a la carte" charts. We had the values right (e.g. `--set hub=x`) and we made a terrible decision, pushed by myself to "be consistent with the other charts"), to scope them under that original chart names (thus `--set cni.hub=x`). 

Now we are adding back a parent chart (but harder, since we are exposing BOTH forms), so we end up doubling our mistake with `--set cni.cni.hub`. That sucks and I don't want that. But I also want to make sure we aren't going to be looking back on decisions we make here and end up with an even bigger mess down the road. For example, why are just doing unscoping for `hub`? Now its even less consistent since I can do `--set cni.hub=foo` but I need to do `--set cni.cni.resources` or something.

Sorry I know there is a lot of back-and-forth here and lack of clarity, but this is an area we have messed up _really bad_ in Istio repeatedly. So we better be really really sure we are doing it right this time.

Concretely as next steps, WDYT of:
* Isolate a global-only fix for ztunnel. This can be https://github.com/istio/istio/pull/51749 or similar, either by undrafting my PR or feel free to take it over, merge it into this one, etc. W/e is easiest
* If we want flat values then propose a solution that:
   * Lets users still do `--set pilot.foo=bar` (when using directly, for back-compat), `--set CHART.foo=bar` (when using nested), and `--set foo=bar` (when using directly).
   * Does not require knowledge of this flattening in each part of the template. I.e we should have `foo: {{.Values.foo}}` or something, not some complex conditionals everywhere.
   * Is consistent, not applying on to some fields and not others
   * I think this could be met  by doing something like `merge .Values .Values.pilot` to flatten it out. Just like the global approach in https://github.com/istio/istio/pull/51749
   * Has buyin from others in env wg like @costinm (ideally before we get too far so we don't waste your time)

Reply:
> Yes, the legacy charts have legacy semantics. Please don't change these without getting a clear understanding of our goals. I very much don't like these changes without clear high level direction + mixed in with "ztunnel not respecting global".
>
> We are in the `cni.cni.hub` position for very similar reasons (caused by me). We went from "Istio is only a parent chart" to "a la carte" charts. We had the values right (e.g. `--set hub=x`) and we made a terrible decision, pushed by myself to "be consistent with the other charts"), to scope them under that original chart names (thus `--set cni.hub=x`).
> 
> Now we are adding back a parent chart (but harder, since we are exposing BOTH forms), so we end up doubling our mistake with `--set cni.cni.hub`. That sucks and I don't want that. But I also want to make sure we aren't going to be looking back on decisions we make here and end up with an even bigger mess down the road. For example, why are just doing unscoping for `hub`? Now its even less consistent since I can do `--set cni.hub=foo` but I need to do `--set cni.cni.resources` or something.
> 
> Sorry I know there is a lot of back-and-forth here and lack of clarity, but this is an area we have messed up _really bad_ in Istio repeatedly. So we better be really really sure we are doing it right this time.

Of course, that's why this PR exists, it's a process, and it requires feedback. The only goal I have here is to make our charts less surprising to someone who isn't familiar with Istio.
 
> Concretely as next steps, WDYT of:
> 
>     * Isolate a global-only fix for ztunnel. This can be [WIP: simplified global management in ztunnel #51749](https://github.com/istio/istio/pull/51749) or similar, either by undrafting my PR or feel free to take it over, merge it into this one, etc. W/e is easiest

:+1: I can do that by going back to a previous iteration of this PR that duplicated image templating in every chart, and split this.

>     * If we want flat values then propose a solution that:
>       
>       * Lets users still do `--set pilot.foo=bar` (when using directly, for back-compat), `--set CHART.foo=bar` (when using nested), and `--set foo=bar` (when using directly).

That is exactly what the most recent revision of this PR does, yes: https://github.com/istio/istio/pull/51517/commits/4054e0bb4f66191de4a7f1e871a5f041ab61fb27

- `--set pilot.hub = bar` -> wins if supplied by the user (but not if encoded in our values/templates - which I think is fine, we own those, as long as the explicit user override works for back compat, we don't have to support customized charts) 
- `--set global.hub = bar` -> works

- `--set hub = bar` -> works (unless global or pilot is set)

- `--set CHART-AS-DEP.hub` -> (nested scenario) works (because Helm does the scoping for us)

>       * Does not require knowledge of this flattening in each part of the template. I.e we should have `foo: {{.Values.foo}}` or something, not some complex conditionals everywhere.

Yes, in this PR we flatten the image values once - in the profile (which is shared by all charts). And we use a templated image helper that doesn't know or care about the unique per-chart scopes for these values, so the actual charts also do not need to know about this.

>       * Is consistent, not applying on to some fields and not others

Yeah this PR does not fix _all_ the scoped vars, just the ones that are A) commonly overriden (image vars) and the ones that _need_ to be fixed to get the shared template to work for all charts.

Unscoping EVERYTHING is probably not something we have to do all at once, but for shared/consistent image templating with globals, we have to do that subset. I think that is fine actually - much like `globals` some values make sense with the scoping, and some do not. I am proposing that _at least for the extremely common image values_ we can and should be consistent. So I would actually prefer to do this incrementally, since I think we can, starting with the most commonly-overridden values.

>       * I think this could be met  by doing something like `merge .Values .Values.pilot` to flatten it out. Just like the global approach in [WIP: simplified global management in ztunnel #51749](https://github.com/istio/istio/pull/51749)
Yeah as mentioned the lack of consistency in image values makes this harder, and we have that lack of consistency because we didn't use a consistent templating func (which this adds) - but we can split this PR, go back to duplicating the templates again, and then improve that in a followup that contains the remaining bits of this PR.

>       * Has buyin from others in env wg like @costinm (ideally before we get too far so we don't waste your time)

My takeaways from @costinm are that he wants

- Consistent behavior
- Unsurprising behavior (for non-Istio helm users)
- Backwards compat
- Strong opinions/controls on what `global` values each chart respects

This PR I believe hits all of those (but I will let him chime in in case I am misrepresenting him)

Reply:
I don't think there is a compelling reason to have a "shared image template", and it seems most of our decisions are coming back to that goal?

> Yeah as mentioned the lack of consistency in image values makes this harder, and we have that lack of consistency because we didn't use a consistent templating func

Disagree with that/don't understand. There is no "image inconsistency", there is just values inconsistency.

IIUC your concern (which is very fair) is that we do not want `cni.cni.tag`. But why would I NOT want that, but DO want `cni.cni.podAnnotations`? Just an example, I am sure there are many many fields.

Reply:
@howardjohn Attempted to do something simpler like https://github.com/istio/istio/pull/51749, by duplicating image templating (or just not doing any imageg templating at all) - unfortunately I think that simply can't work correctly in all scenarios.

Either it results in `global` being silently ignored, because the keys differ between charts, or it results in surprising precedence problems that will show up in nesting (non-global CLI override loses to global CLI override, which means you can't do `--set global.hub=bar --set <subchart>.hub=foo` and have <subchart> get `foo`) 

This is largely a side effect of the existing charts being nonstandard in terms of variable paths (which this PR fixes) and the extant profile munging requiring us to take extra care with precedence (which this PR accounts for).

The approach in this PR is the only one I can get to work correctly (and by "correctly" I mean mimic vanilla Helm) in all cases:

(taking `pilot` as an example because it has the scoping, which is the most complex case):

- explicitly supplied by the user via CLI always wins (global or not-global) (e.g. regardless of the `values.yaml`, `--set pilot.hub=foo` or `--set global.hub=foo` should always work for all charts, both should beat the stamped defaults in the chart values)

- if both global and local supplied by the user via CLI, the more specifically scoped one always wins

There are several interlocking things here which make this hard to do simply or incrementally, unfortunately. We can't fix one without breaking something else, or causing an unexpected effect in another spot - so they have to be fixed together, like in this PR. Anything simpler just doesn't work correctly in all cases.

Reply:
> Disagree with that/don't understand. There is no "image inconsistency", there is just values inconsistency.

That's correct, I am just focused on the image values specifically here, but generally that is true.

Reply:
I would _much_ rather have everything be bad (where bad == `cni.cni`) than to be inconsistently bad. its far easier to grok and workaround consistently-weird things than inconsistently weird things

Reply:
It is already inconsistently bad across charts. This makes image values (which are among the ones most likely to have collision and scoping issues)  consistently good across charts. It leaves some of the charts inconsistently bad within themselves in terms of the scoping.


This is an incremental improvement, not a complete improvement, that is correct.

Reply:
Its consistent _across_ charts, but consistent _within_ charts. Now its inconsistent across both. It is an incremental change that makes the state worse, and is not incrementally moving to the end state that meets https://github.com/istio/istio/pull/51517#discussion_r1657706591

Reply:
Sorry, I'm probably not grokking.

Before, it was impossible to set commonly-overriden image params consistently across ambient charts.

With this PR it is. This PR doesn't fix _all_ params for _all_ charts, but could be extended to do that in followups. The image params are the ones most likely to need overriding, and the ones that need to properly handle nesting the most. There are others too, but it makes sense to start here, and start small.

With this PR:

- it is possible to set image properties the same way for all charts, with override precedence that works like Regular Helm (and it fixes tests that were doing bad/inconsistent overrides)
- additionally this adds templating chart image fields consistently in the same way across all ambient charts (we were not before, this was a source of bugs)
- additionally supports a fixed subset of global values across all charts in a way that works correctly across all charts, and when composed.


Also this PR correctly handles the following permutations of image value overrides correctly

- `--set pilot.hub/tag/variant/pullPolicy = bar` -> legacy, becomes a problem if we compose
- `--set global.hub/tag/variant/pullPolicy = bar`
`- --set hub/tag/variant/pullPolicy = bar`
`--set CHART-AS-DEP.hub/tag/variant/pullPolicy`


If you test and find this PR doesn't meet those reqs, or there is a simpler and more consistent way to meet those reqs that fixes the test and the UX, we can do that - this is as close to normative, unsurprising Helm we can get and still support all the Istio-specific opinions we currently support and want to support, AFAICT. 

Either way this is currently passing CI and working correctly ATM, assuming https://github.com/istio/release-builder/pull/1871 goes in - it's here, we can take it or not. If we don't take it, we still have (several) usage and UX problems that will have to be dealt with in similar ways.